File: apps/api-rs/src/document/shared.rs
use super::*;
pub(super) fn attach_table_cell_anchors(
    doc_id: Uuid,
    parse_job_id: Uuid,
    format: &str,
    blocks: &mut [ParsedBlock],
    tables: &[ParsedTable],
    anchors: &mut Vec<SourceAnchor>,
) {
    for block in blocks.iter_mut().filter(|block| block.table_id.is_some()) {
        if block
            .anchor_ids
            .iter()
            .any(|anchor_id| anchors.iter().any(|anchor| anchor.anchor_id == *anchor_id))
        {
            continue;
        }
        let Some(table_id) = block.table_id else {
            continue;
        };
        let Some(table) = tables.iter().find(|table| table.table_id == table_id) else {
            continue;
        };
        let cell_range = table_cell_range(table);
        let source_ref = json!({
            "format": format,
            "table_index": table.table_index,
            "table_id": table.table_id,
            "kind": "table_cell_range",
            "cell_range": {
                "row_start": cell_range.row_start,
                "row_end": cell_range.row_end,
                "col_start": cell_range.col_start,
                "col_end": cell_range.col_end,
            }
        });
        let anchor = SourceAnchor::table_cell_range(
            doc_id,
            parse_job_id,
            Uuid::nil(),
            format,
            block.block_id,
            table.table_id,
            table.page_start.or(block.page_start),
            table.slide_index.or(block.slide_index),
            cell_range,
            source_ref,
            &table.markdown,
        );
        block.anchor_ids.push(anchor.anchor_id);
        anchors.push(anchor);
    }
}

fn table_cell_range(table: &ParsedTable) -> CellRange {
    let row_end = table
        .cells
        .iter()
        .map(|cell| cell.row_index + cell.rowspan.max(1) - 1)
        .max()
        .unwrap_or(0);
    let col_end = table
        .cells
        .iter()
        .map(|cell| cell.col_index + cell.colspan.max(1) - 1)
        .max()
        .unwrap_or(0);
    CellRange {
        row_start: 0,
        row_end,
        col_start: 0,
        col_end,
    }
}

pub(super) fn finalize_parsed(
    doc_id: Uuid,
    parse_job_id: Uuid,
    file_type: &str,
    title: &str,
    pages: Option<i32>,
    blocks: Vec<ParsedBlock>,
    tables: Vec<ParsedTable>,
    anchors: Vec<SourceAnchor>,
) -> ParsedDocument {
    ParsedDocument {
        doc_id,
        parse_job_id,
        file_type: file_type.to_string(),
        title: title.to_string(),
        pages,
        blocks,
        tables,
        anchors,
        warnings: vec![],
        quality_score: 0.0,
    }
}

pub(super) fn parse_xml_table(
    tbl: roxmltree::Node,
    _doc_id: Uuid,
    _parse_job_id: Uuid,
    block_id: Uuid,
    table_index: i32,
    heading_path: Vec<String>,
    page_start: Option<i32>,
    page_end: Option<i32>,
    slide_index: Option<i32>,
    format: &str,
) -> ParsedTable {
    let table_id = Uuid::new_v4();
    let row_nodes: Vec<_> = tbl
        .children()
        .filter(|n| n.is_element() && n.tag_name().name() == "tr")
        .collect();
    let mut rows = Vec::new();
    let mut cells: Vec<ParsedTableCell> = Vec::new();
    let mut vertical_merges = std::collections::BTreeMap::<i32, usize>::new();
    for (row_idx, row) in row_nodes.iter().enumerate() {
        let cell_nodes: Vec<_> = row
            .children()
            .filter(|n| n.is_element() && n.tag_name().name() == "tc")
            .collect();
        let mut row_values = Vec::new();
        let mut col_idx = 0_i32;
        for cell in &cell_nodes {
            let text = collect_text(*cell).trim().to_string();
            let colspan = attr_i32(*cell, "gridSpan")
                .or_else(|| {
                    cell.descendants()
                        .find(|n| n.tag_name().name() == "gridSpan")
                        .and_then(|n| attr_i32(n, "val"))
                })
                .unwrap_or(1)
                .max(1);
            let vertical_merge = cell
                .descendants()
                .find(|node| node.tag_name().name() == "vMerge");
            let continues_vertical_merge = vertical_merge
                .and_then(|node| attr(node, "val"))
                .map(|value| value != "restart")
                .unwrap_or(vertical_merge.is_some());
            if continues_vertical_merge {
                if let Some(origin_index) = vertical_merges.get(&col_idx).copied() {
                    cells[origin_index].rowspan += 1;
                }
                row_values.extend(std::iter::repeat_n(String::new(), colspan as usize));
                col_idx += colspan;
                continue;
            }

            row_values.push(text.clone());
            row_values.extend(std::iter::repeat_n(
                String::new(),
                colspan.saturating_sub(1) as usize,
            ));
            let cell_index = cells.len();
            cells.push(ParsedTableCell {
                cell_id: Uuid::new_v4(),
                row_index: row_idx as i32,
                col_index: col_idx,
                rowspan: attr_i32(*cell, "rowSpan").unwrap_or(1).max(1),
                colspan,
                normalized_text: normalize_space(&text),
                text,
                is_header: row_idx == 0,
                data_type: "text".to_string(),
                bbox: None,
                style: json!({}),
                source_ref: json!({"format": format, "table_index": table_index, "row": row_idx, "col": col_idx}),
            });
            for column in col_idx..col_idx + colspan {
                if vertical_merge
                    .and_then(|node| attr(node, "val"))
                    .is_some_and(|value| value == "restart")
                {
                    vertical_merges.insert(column, cell_index);
                } else {
                    vertical_merges.remove(&column);
                }
            }
            col_idx += colspan;
        }
        if !row_values.iter().all(|v| v.is_empty()) {
            rows.push(row_values);
        }
    }
    let headers = rows.first().cloned().unwrap_or_default();
    let markdown = table_markdown(&headers, rows.get(1..).unwrap_or(&[]));
    let total_slots = rows.iter().map(Vec::len).sum::<usize>();
    let empty_slots = rows
        .iter()
        .flat_map(|row| row.iter())
        .filter(|value| value.trim().is_empty())
        .count();
    let empty_cell_ratio = if total_slots == 0 {
        1.0
    } else {
        empty_slots as f64 / total_slots as f64
    };
    ParsedTable {
        table_id,
        block_id,
        table_index,
        title: heading_path.last().cloned(),
        heading_path,
        page_start,
        page_end,
        slide_index,
        headers,
        rows: rows.get(1..).unwrap_or(&[]).to_vec(),
        cells,
        markdown,
        quality: json!({
            "header_confidence": if rows.is_empty() { 0.0 } else { 0.9 },
            "grid_confidence": 0.95,
            "empty_cell_ratio": empty_cell_ratio,
            "warnings": []
        }),
        source_ref: json!({"format": format, "table_index": table_index}),
    }
}

pub(super) fn zip_entry_names(bytes: &[u8]) -> Result<Vec<String>> {
    let mut archive = open_zip(bytes)?;
    Ok((0..archive.len())
        .filter_map(|i| archive.by_index(i).ok().map(|f| f.name().to_string()))
        .collect())
}

pub(super) fn open_zip(bytes: &[u8]) -> Result<ZipArchive<Cursor<&[u8]>>> {
    let mut archive = ZipArchive::new(Cursor::new(bytes)).context("invalid_zip_container")?;
    validate_office_zip(&mut archive)?;
    Ok(archive)
}

pub(super) fn read_zip_text(archive: &mut ZipArchive<Cursor<&[u8]>>, name: &str) -> Result<String> {
    let mut file = archive
        .by_name(name)
        .with_context(|| format!("missing_zip_entry:{name}"))?;
    if file.size() > MAX_OFFICE_XML_BYTES {
        bail!(
            "zip_xml_entry_too_large:{name}:{}>{}",
            file.size(),
            MAX_OFFICE_XML_BYTES
        );
    }
    let mut text = String::new();
    file.read_to_string(&mut text)?;
    Ok(text)
}

pub(super) fn validate_xml_nesting(xml: &str) -> Result<()> {
    use quick_xml::events::Event;

    let mut reader = quick_xml::Reader::from_str(xml);
    let mut depth = 0usize;
    loop {
        match reader.read_event() {
            Ok(Event::Start(_)) => {
                depth += 1;
                if depth > MAX_OFFICE_XML_DEPTH {
                    bail!("office_xml_nesting_exceeded:{depth}>{MAX_OFFICE_XML_DEPTH}");
                }
            }
            Ok(Event::End(_)) => {
                depth = depth
                    .checked_sub(1)
                    .ok_or_else(|| anyhow!("invalid_office_xml_nesting"))?;
            }
            Ok(Event::Eof) => break,
            Ok(_) => {}
            Err(error) => return Err(anyhow!("invalid_office_xml:{error}")),
        }
    }
    if depth != 0 {
        bail!("invalid_office_xml_nesting");
    }
    Ok(())
}

fn validate_office_zip(archive: &mut ZipArchive<Cursor<&[u8]>>) -> Result<()> {
    if archive.len() > MAX_OFFICE_ZIP_ENTRIES {
        bail!(
            "zip_entry_count_exceeded:{}>{}",
            archive.len(),
            MAX_OFFICE_ZIP_ENTRIES
        );
    }

    let mut total_uncompressed = 0_u64;
    for index in 0..archive.len() {
        let file = archive
            .by_index(index)
            .with_context(|| format!("invalid_zip_entry:{index}"))?;
        let name = file.name().to_string();
        if !is_safe_zip_entry_name(&name) {
            bail!("zip_entry_name_unsafe:{name}");
        }

        let uncompressed = file.size();
        let compressed = file.compressed_size();
        if uncompressed > MAX_OFFICE_ENTRY_BYTES {
            bail!("zip_entry_too_large:{name}:{uncompressed}>{MAX_OFFICE_ENTRY_BYTES}");
        }
        total_uncompressed = total_uncompressed
            .checked_add(uncompressed)
            .ok_or_else(|| anyhow!("zip_uncompressed_size_overflow"))?;
        if total_uncompressed > MAX_OFFICE_UNCOMPRESSED_BYTES {
            bail!(
                "zip_uncompressed_size_exceeded:{total_uncompressed}>{MAX_OFFICE_UNCOMPRESSED_BYTES}"
            );
        }
        if uncompressed > 0 && compressed == 0 {
            bail!("zip_entry_invalid_compressed_size:{name}");
        }
        if compressed > 0 && uncompressed > compressed.saturating_mul(MAX_OFFICE_COMPRESSION_RATIO)
        {
            bail!(
                "zip_compression_ratio_exceeded:{name}:{uncompressed}/{compressed}>{MAX_OFFICE_COMPRESSION_RATIO}"
            );
        }
    }

    Ok(())
}

fn is_safe_zip_entry_name(name: &str) -> bool {
    if name.trim().is_empty()
        || name.starts_with('/')
        || name.starts_with('\\')
        || name.contains('\0')
        || name.contains(':')
    {
        return false;
    }
    let normalized = name.replace('\\', "/");
    normalized
        .trim_end_matches('/')
        .split('/')
        .all(|part| !part.is_empty() && part != "." && part != "..")
}

pub(super) fn collect_text(node: roxmltree::Node) -> String {
    let mut text = String::new();
    for n in node.descendants() {
        if n.is_text() {
            text.push_str(n.text().unwrap_or_default());
        } else if n.is_element() && matches!(n.tag_name().name(), "tab") {
            text.push('\t');
        } else if n.is_element() && matches!(n.tag_name().name(), "br" | "cr") {
            text.push('\n');
        }
    }
    normalize_space(&text)
}

pub(super) fn attr<'a>(node: roxmltree::Node<'a, 'a>, local_name: &str) -> Option<&'a str> {
    node.attributes()
        .find(|a| a.name() == local_name)
        .map(|a| a.value())
}

pub(super) fn attr_i32(node: roxmltree::Node, local_name: &str) -> Option<i32> {
    attr(node, local_name).and_then(|v| v.parse().ok())
}

pub(super) fn normalize_space(value: &str) -> String {
    value.split_whitespace().collect::<Vec<_>>().join(" ")
}

pub(super) fn table_markdown(headers: &[String], rows: &[Vec<String>]) -> String {
    if headers.is_empty() {
        return rows
            .iter()
            .map(|row| row.join(" | "))
            .collect::<Vec<_>>()
            .join("\n");
    }
    let mut out = String::new();
    out.push('|');
    out.push_str(
        &headers
            .iter()
            .map(|h| escape_table_cell(h))
            .collect::<Vec<_>>()
            .join("|"),
    );
    out.push_str("|\n|");
    out.push_str(&headers.iter().map(|_| "---").collect::<Vec<_>>().join("|"));
    out.push_str("|\n");
    for row in rows {
        out.push('|');
        out.push_str(
            &row.iter()
                .map(|v| escape_table_cell(v))
                .collect::<Vec<_>>()
                .join("|"),
        );
        out.push_str("|\n");
    }
    out
}

fn escape_table_cell(value: &str) -> String {
    value.replace('|', "\\|").replace('\n', "<br>")
}

File: apps/api-bun/src/document/shared.ts
// 移植自 apps/api-rs/src/document/shared.rs —— 行为对齐 Rust 原版，错误信息保持一致

// 共享逻辑：zip 安全校验与读取、XML 表格解析、表格锚点、文本收集、markdown 表格渲染

import JSZip from 'jszip';
import type { JSZipObject } from 'jszip';
import { newUuid } from '../infra/uuid.ts';
import type { CellRange, SourceAnchor } from '../models/source_anchor.ts';
import { sourceAnchorTableCellRange } from '../models/source_anchor.ts';
import {
  MAX_OFFICE_COMPRESSION_RATIO,
  MAX_OFFICE_ENTRY_BYTES,
  MAX_OFFICE_UNCOMPRESSED_BYTES,
  MAX_OFFICE_XML_BYTES,
  MAX_OFFICE_ZIP_ENTRIES,
  NIL_UUID,
  type ParsedBlock,
  type ParsedDocument,
  type ParsedTable,
  type ParsedTableCell,
} from './types.ts';
import { attrI32, childElements, descendants, attr, type XmlElement } from './xml.ts';
import { splitWhitespace } from './text_utils.ts';

export function normalizeSpace(value: string): string {
  return splitWhitespace(value).join(' ');
}

/** 对齐 Rust collect_text：后代文本拼接，tab -> \t，br/cr -> \n，最后 normalize_space */
export function collectText(node: XmlElement): string {
  let text = '';
  for (const ref of descendants(node)) {
    if (ref.kind === 'text') {
      text += ref.text;
    } else if (ref.element.name === 'tab') {
      text += '\t';
    } else if (ref.element.name === 'br' || ref.element.name === 'cr') {
      text += '\n';
    }
  }
  return normalizeSpace(text);
}

export function finalizeParsed(
  docId: string,
  parseJobId: string,
  fileType: string,
  title: string,
  pages: number | null,
  blocks: ParsedBlock[],
  tables: ParsedTable[],
  anchors: SourceAnchor[],
): ParsedDocument {
  return {
    doc_id: docId,
    parse_job_id: parseJobId,
    file_type: fileType,
    title,
    pages,
    blocks,
    tables,
    anchors,
    warnings: [],
    quality_score: 0.0,
  };
}

/** 对齐 Rust attach_table_cell_anchors：为表格块补充 table_cell_range 锚点 */
export function attachTableCellAnchors(
  docId: string,
  parseJobId: string,
  format: string,
  blocks: ParsedBlock[],
  tables: ParsedTable[],
  anchors: SourceAnchor[],
): void {
  for (const block of blocks) {
    if (block.table_id === null) continue;
    if (block.anchor_ids.some((anchorId) => anchors.some((anchor) => anchor.anchor_id === anchorId))) {
      continue;
    }
    const table = tables.find((candidate) => candidate.table_id === block.table_id);
    if (!table) continue;
    const cellRange = tableCellRange(table);
    const sourceRef = {
      format,
      table_index: table.table_index,
      table_id: table.table_id,
      kind: 'table_cell_range',
      cell_range: {
        row_start: cellRange.row_start,
        row_end: cellRange.row_end,
        col_start: cellRange.col_start,
        col_end: cellRange.col_end,
      },
    };
    const anchor = sourceAnchorTableCellRange(
      docId,
      parseJobId,
      NIL_UUID,
      format,
      block.block_id,
      table.table_id,
      table.page_start ?? block.page_start,
      table.slide_index ?? block.slide_index,
      cellRange,
      sourceRef,
      table.markdown,
    );
    block.anchor_ids.push(anchor.anchor_id);
    anchors.push(anchor);
  }
}

function tableCellRange(table: ParsedTable): CellRange {
  let rowEnd = 0;
  let colEnd = 0;
  for (const cell of table.cells) {
    rowEnd = Math.max(rowEnd, cell.row_index + Math.max(cell.rowspan, 1) - 1);
    colEnd = Math.max(colEnd, cell.col_index + Math.max(cell.colspan, 1) - 1);
  }
  return { row_start: 0, row_end: rowEnd, col_start: 0, col_end: colEnd };
}

/** 对齐 Rust parse_xml_table：w:tbl / a:tbl 网格解析（gridSpan / vMerge / rowSpan） */
export function parseXmlTable(
  tbl: XmlElement,
  _docId: string,
  _parseJobId: string,
  blockId: string,
  tableIndex: number,
  headingPath: string[],
  pageStart: number | null,
  pageEnd: number | null,
  slideIndex: number | null,
  format: string,
): ParsedTable {
  const tableId = newUuid();
  const rowNodes = childElements(tbl).filter((node) => node.name === 'tr');
  const rows: string[][] = [];
  const cells: ParsedTableCell[] = [];
  const verticalMerges = new Map<number, number>();

  for (const [rowIdx, row] of rowNodes.entries()) {
    const cellNodes = childElements(row).filter((node) => node.name === 'tc');
    const rowValues: string[] = [];
    let colIdx = 0;
    for (const cell of cellNodes) {
      const text = collectText(cell).trim();
      const gridSpanEl = descendants(cell).find(
        (ref) => ref.kind === 'element' && ref.element.name === 'gridSpan',
      );
      const colspan = Math.max(
        attrI32(cell, 'gridSpan')
          ?? (gridSpanEl?.kind === 'element' ? attrI32(gridSpanEl.element, 'val') : undefined)
          ?? 1,
        1,
      );
      const vMergeEl = descendants(cell).find(
        (ref) => ref.kind === 'element' && ref.element.name === 'vMerge',
      );
      const vMerge = vMergeEl?.kind === 'element' ? vMergeEl.element : undefined;
      const vMergeVal = vMerge ? attr(vMerge, 'val') : undefined;
      const continuesVerticalMerge = vMerge
        ? (vMergeVal !== undefined ? vMergeVal !== 'restart' : true)
        : false;
      if (continuesVerticalMerge) {
        const originIndex = verticalMerges.get(colIdx);
        if (originIndex !== undefined) {
          const origin = cells[originIndex];
          if (origin) origin.rowspan += 1;
        }
        for (let i = 0; i < colspan; i += 1) rowValues.push('');
        colIdx += colspan;
        continue;
      }

      rowValues.push(text);
      for (let i = 0; i < Math.max(colspan - 1, 0); i += 1) rowValues.push('');
      const cellIndex = cells.length;
      cells.push({
        cell_id: newUuid(),
        row_index: rowIdx,
        col_index: colIdx,
        rowspan: Math.max(attrI32(cell, 'rowSpan') ?? 1, 1),
        colspan,
        normalized_text: normalizeSpace(text),
        text,
        is_header: rowIdx === 0,
        data_type: 'text',
        bbox: null,
        style: {},
        source_ref: { format, table_index: tableIndex, row: rowIdx, col: colIdx },
      });
      for (let column = colIdx; column < colIdx + colspan; column += 1) {
        if (vMergeVal === 'restart') {
          verticalMerges.set(column, cellIndex);
        } else {
          verticalMerges.delete(column);
        }
      }
      colIdx += colspan;
    }
    if (!rowValues.every((value) => value.length === 0)) {
      rows.push(rowValues);
    }
  }
  const headers = rows.length > 0 ? (rows[0] ?? []) : [];
  const body = rows.slice(1);
  const markdown = tableMarkdown(headers, body);
  const totalSlots = rows.reduce((sum, row) => sum + row.length, 0);
  const emptySlots = rows.reduce(
    (sum, row) => sum + row.filter((value) => value.trim().length === 0).length,
    0,
  );
  const emptyCellRatio = totalSlots === 0 ? 1.0 : emptySlots / totalSlots;
  return {
    table_id: tableId,
    block_id: blockId,
    table_index: tableIndex,
    title: headingPath.length > 0 ? headingPath[headingPath.length - 1]! : null,
    heading_path: headingPath,
    page_start: pageStart,
    page_end: pageEnd,
    slide_index: slideIndex,
    headers,
    rows: body,
    cells,
    markdown,
    quality: {
      header_confidence: rows.length === 0 ? 0.0 : 0.9,
      grid_confidence: 0.95,
      empty_cell_ratio: emptyCellRatio,
      warnings: [],
    },
    source_ref: { format, table_index: tableIndex },
  };
}

// ---------------- zip 安全校验与读取（对齐 Rust shared::{open_zip, read_zip_text, ...}） ----------------

interface ZipEntrySize {
  uncompressedSize: number;
  compressedSize: number;
}

/** jszip 未在类型中暴露 central directory 尺寸元数据，运行时校验后显式使用 */
function zipEntrySize(file: JSZipObject): ZipEntrySize {
  const internal = file as unknown as { _data?: { uncompressedSize?: number; compressedSize?: number } };
  const uncompressedSize = internal._data?.uncompressedSize;
  const compressedSize = internal._data?.compressedSize;
  if (typeof uncompressedSize !== 'number' || typeof compressedSize !== 'number') {
    throw new Error('zip_entry_meta_missing:' + file.name);
  }
  return { uncompressedSize, compressedSize };
}

export async function openZip(bytes: Uint8Array): Promise<JSZip> {
  let zip: JSZip;
  try {
    zip = await JSZip.loadAsync(bytes);
  } catch {
    throw new Error('invalid_zip_container');
  }
  validateOfficeZip(zip);
  return zip;
}

export async function zipEntryNames(bytes: Uint8Array): Promise<string[]> {
  const zip = await openZip(bytes);
  return Object.values(zip.files).map((file) => file.name);
}

export async function readZipText(zip: JSZip, name: string): Promise<string> {
  const file = zip.file(name);
  if (!file) throw new Error('missing_zip_entry:' + name);
  const { uncompressedSize } = zipEntrySize(file);
  if (uncompressedSize > MAX_OFFICE_XML_BYTES) {
    throw new Error('zip_xml_entry_too_large:' + name + ':' + uncompressedSize + '>' + MAX_OFFICE_XML_BYTES);
  }
  const raw = await file.async('uint8array');
  try {
    return new TextDecoder('utf-8', { fatal: true }).decode(raw);
  } catch {
    throw new Error('invalid_zip_entry_encoding:' + name);
  }
}

function validateOfficeZip(zip: JSZip): void {
  const entries = Object.values(zip.files);
  if (entries.length > MAX_OFFICE_ZIP_ENTRIES) {
    throw new Error('zip_entry_count_exceeded:' + entries.length + '>' + MAX_OFFICE_ZIP_ENTRIES);
  }
  let totalUncompressed = 0;
  for (const file of entries) {
    if (file.dir) continue;
    const name = file.name;
    if (!isSafeZipEntryName(name)) {
      throw new Error('zip_entry_name_unsafe:' + name);
    }
    const { uncompressedSize, compressedSize } = zipEntrySize(file);
    if (uncompressedSize > MAX_OFFICE_ENTRY_BYTES) {
      throw new Error('zip_entry_too_large:' + name + ':' + uncompressedSize + '>' + MAX_OFFICE_ENTRY_BYTES);
    }
    totalUncompressed += uncompressedSize;
    if (!Number.isSafeInteger(totalUncompressed)) {
      throw new Error('zip_uncompressed_size_overflow');
    }
    if (totalUncompressed > MAX_OFFICE_UNCOMPRESSED_BYTES) {
      throw new Error('zip_uncompressed_size_exceeded:' + totalUncompressed + '>' + MAX_OFFICE_UNCOMPRESSED_BYTES);
    }
    if (uncompressedSize > 0 && compressedSize === 0) {
      throw new Error('zip_entry_invalid_compressed_size:' + name);
    }
    if (compressedSize > 0 && uncompressedSize > compressedSize * MAX_OFFICE_COMPRESSION_RATIO) {
      throw new Error(
        'zip_compression_ratio_exceeded:' + name + ':' + uncompressedSize + '/' + compressedSize + '>' + MAX_OFFICE_COMPRESSION_RATIO,
      );
    }
  }
}

function isSafeZipEntryName(name: string): boolean {
  if (name.trim().length === 0
    || name.startsWith('/')
    || name.startsWith('\\')
    || name.includes('\0')
    || name.includes(':')) {
    return false;
  }
  const normalized = name.replace(/\\/g, '/');
  const trimmed = normalized.endsWith('/') ? normalized.slice(0, -1) : normalized;
  return trimmed.split('/').every((part) => part.length > 0 && part !== '.' && part !== '..');
}

// ---------------- markdown 表格渲染 ----------------

export function tableMarkdown(headers: string[], rows: string[][]): string {
  if (headers.length === 0) {
    return rows.map((row) => row.join(' | ')).join('\n');
  }
  let out = '';
  out += '|' + headers.map((header) => escapeTableCell(header)).join('|') + '|\n';
  out += '|' + headers.map(() => '---').join('|') + '|\n';
  for (const row of rows) {
    out += '|' + row.map((value) => escapeTableCell(value)).join('|') + '|\n';
  }
  return out;
}

function escapeTableCell(value: string): string {
  return value.replace(/\|/g, '\\|').replace(/\n/g, '<br>');
}
